File: src/main/java/core/procedures/strings/StringComparison.java
package core.procedures.strings;

import core.procedures.AFn;
import core.procedures.FnArgsBuilder;

import java.util.function.BiPredicate;

public final class StringComparison extends AFn {

  @Override
  public boolean isPure() {
    return true;
  }

  public static final StringComparison STRING_EQ          = new StringComparison("string=?",     String::equals);
  public static final StringComparison STRING_EQ_CI       = new StringComparison("string-ci=?",  String::equalsIgnoreCase);
  public static final StringComparison STRING_LE          = new StringComparison("string<?",     (arg1, arg2) -> arg1.compareTo(arg2) < 0);
  public static final StringComparison STRING_GR          = new StringComparison("string>?",     (arg1, arg2) -> arg1.compareTo(arg2) > 0);
  public static final StringComparison STRING_LE_OR_EQ    = new StringComparison("string<=?",    (arg1, arg2) -> arg1.compareTo(arg2) <= 0);
  public static final StringComparison STRING_GR_OR_EQ    = new StringComparison("string>=?",    (arg1, arg2) -> arg1.compareTo(arg2) >= 0);
  public static final StringComparison STRING_LE_CI       = new StringComparison("string-ci<?",  (arg1, arg2) -> arg1.toLowerCase().compareTo(arg2.toLowerCase()) < 0);
  public static final StringComparison STRING_GR_CI       = new StringComparison("string-ci>?",  (arg1, arg2) -> arg1.toLowerCase().compareTo(arg2.toLowerCase()) > 0);
  public static final StringComparison STRING_LE_OR_EQ_CI = new StringComparison("string-ci<=?", (arg1, arg2) -> arg1.toLowerCase().compareTo(arg2.toLowerCase()) <= 0);
  public static final StringComparison STRING_GR_OR_EQ_CI = new StringComparison("string-ci>=?", (arg1, arg2) -> arg1.toLowerCase().compareTo(arg2.toLowerCase()) >= 0);

  private final String name;
  private final BiPredicate<String, String> predicate;

  private StringComparison(String name, BiPredicate<String, String> predicate) {
    super(new FnArgsBuilder().restArgsType(String.class));
    this.name = name;
    this.predicate = predicate;
  }

  @Override
  public Boolean apply(Object... args) {
    if (args.length < 2) {
      return Boolean.TRUE;
    }
    for (int i = 0; i < args.length - 1; i++) {
      if ((!predicate.test(args[i].toString(), args[i + 1].toString()))) {
        return Boolean.FALSE;
      }
    }
    return Boolean.TRUE;
  }

  @Override
  public String getName() {
    return name;
  }
}

File: src/main/java/core/procedures/strings/StringComparison.kt
package core.procedures.strings

import core.procedures.AFn

class StringComparison private constructor(override val name: String,
                                           inline private val predicate: (String, String) -> Boolean) :
        AFn<Any?, Boolean>(isPure = true, restArgsType = CharSequence::class.java) {

    companion object {
        val STRING_EQ          = StringComparison("string=?",     { f, s -> f == s })
        val STRING_EQ_CI       = StringComparison("string-ci=?",  { f, s -> f.equals(s, true) })
        val STRING_LE          = StringComparison("string<?",     { f, s -> f <  s })
        val STRING_GR          = StringComparison("string>?",     { f, s -> f >  s })
        val STRING_LE_OR_EQ    = StringComparison("string<=?",    { f, s -> f <= s })
        val STRING_GR_OR_EQ    = StringComparison("string>=?",    { f, s -> f >= s })
        val STRING_LE_CI       = StringComparison("string-ci<?",  { f, s -> f.toLowerCase() <  s.toLowerCase() })
        val STRING_GR_CI       = StringComparison("string-ci>?",  { f, s -> f.toLowerCase() >  s.toLowerCase() })
        val STRING_LE_OR_EQ_CI = StringComparison("string-ci<=?", { f, s -> f.toLowerCase() <= s.toLowerCase() })
        val STRING_GR_OR_EQ_CI = StringComparison("string-ci>=?", { f, s -> f.toLowerCase() >= s.toLowerCase() })
    }

    override operator fun invoke(args: Array<out Any?>) = when {
        args.size < 2  -> true
        args.size == 2 -> predicate(args[0].toString(), args[1].toString())
        else           -> (0..args.size - 2).all { predicate(args[it].toString(), args[it + 1].toString()) }
    }
}
